### PR TITLE
Make the build pass with QuickCheck 2.15

### LINKED ISSUE #883: `hspec` version bounds permit QuickCheck 2.15 - breaking the build
Hi,

we recently bumped QuickCheck to 2.15 with one public breaking change: properties are no longer implicitly wrapped in `once`. However, that also appears to have broken the build of `hspec` as you rely on the internal representation of `State`. It's not clear that `State` is internal from the definitions in `QuickCheck` so that's mostly `QuickCheck`'s fault not yours. Anyway, let's try to fix the problem as fast as possible.

I checked out the repo and the following diff fixes your build:
```
diff --git a/hspec-core/src/Test/Hspec/Core/QuickCheck/Util.hs b/hspec-core/src/Test/Hspec/Core/QuickCheck/Util.hs
index c847c391..0705b254 100644
--- a/hspec-core/src/Test/Hspec/Core/QuickCheck/Util.hs
+++ b/hspec-core/src/Test/Hspec/Core/QuickCheck/Util.hs
@@ -130,8 +130,9 @@ showTestCount success discarded = QC.showTestCount state
       terminal                  = undefined
     , maxSuccessTests           = undefined
     , maxDiscardedRatio         = undefined
+    , maxTestSize               = 100
+    , replayStartSize           = Nothing
     , coverageConfidence        = undefined
-    , computeSize               = undefined
     , numTotMaxShrinks          = 0
     , numSuccessTests           = success
     , numDiscardedTests         = discarded
diff --git a/hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs b/hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs
index 831b2e43..4ed68150 100644
--- a/hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs
+++ b/hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs
@@ -117,7 +117,7 @@ spec = do
               , "Passed:"
               , "23"
               , ""
-              , "+++ OK, passed 2 tests."
+              , "+++ OK, passed 2 tests (0% is 5)."
               , ""
               , "Only 0% is 5, but expected 10%"
               ]
@@ -146,7 +146,7 @@ spec = do
             QuickCheckResult 800 "" (QuickCheckFailure failure)
 
         it "includes verbose output" $ do
-          let info = intercalate "\n\n" (replicate 799 "Passed:")
+          let info = intercalate "\n\n" (replicate 800 "Passed:")
           parseQuickCheckResult <$> qc (verbose . p) `shouldReturn`
             QuickCheckResult 800 info (QuickCheckFailure failure)
 
diff --git a/hspec-core/test/Test/Hspec/Core/RunnerSpec.hs b/hspec-core/test/Test/Hspec/Core/RunnerSpec.hs
index 29ec39c2..bbdb8e3e 100644
--- a/hspec-core/test/Test/Hspec/Core/RunnerSpec.hs
+++ b/hspec-core/test/Test/Hspec/Core/RunnerSpec.hs
@@ -800,7 +800,7 @@ spec = do
       let
         setChatty value args = args { chatty = value }
         withChatty value = hspecCapture [] .  H.modifyArgs (setChatty value) $ do
-          H.it "foo" $ property True
+          H.it "foo" $ once $ property True
 
       context "when True" $ do
         it "includes informational output" $ do
```

Now, there are three things going on here:
1. The new `State` representation
2. The change to the output of things that haven't been covered
3. The `once` thing.

In reverse order fixing (3) is just a case of applying the diff above - it won't break backwards compatibility in `hspec`. Fixing (2) probably needs a version pragma on the QuickCheck version (sorry about that) and fixing (1) is the interesting case. Because `State` is, albeit informally, considered internal to QuickCheck this is a bit thorny. On the one hand we could introduce version bound pragmas here too but that just kicks the can down the road. Another way of working around this is to use `withState :: Args -> (State -> IO a) -> IO a` from `Test.QuickCheck.Test` to do what you're currently doing by creating the `undefined`-y state in `Utils.hs`.

What do you guys think?

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This closes #883

### CODE REVIEW

**Comment by andreasabel** [comment]
@MaximilianAlgehed A trick to get the workflows to run without waiting for maintainer approval is to make a PR against your own fork (`Quviq:main`).  Then the workflows run in your fork.  (I do this routinely...)

File: hspec-core/test/Test/Hspec/Core/RunnerSpec.hs
Comment (sol) [inline hspec-core/test/Test/Hspec/Core/RunnerSpec.hs:803]: I guess this works as far as Hspec is concerned. I'm still puzzled how this might affect QuickCheck users in general. E.g. a user might experience a slowdown for an existing test suite, without understanding what introduced that slowdown.

File: hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs
Comment (sol) [inline hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs:156]: 👍

**Review by sol — COMMENTED** [review]
Thanks a lot for helping with this 🙏

**Comment by sol** [comment]
@MaximilianAlgehed https://github.com/hspec/hspec/pull/885 adds an upper bound for QuickCheck. Can you rebase and relax that bound accordingly?

File: version.yaml
Comment (MaximilianAlgehed) [reply version.yaml:1]: I didn't (didn't know it existed, sorry!). I'll take that for a swing now.

**Comment by MaximilianAlgehed** [comment]
I've rebased and relaxed the bound to `<2.16`

File: version.yaml
Comment (MaximilianAlgehed) [reply version.yaml:1]: I went over and re-did this with `util/release` and as far as I can tell it's all good now. Please have a quick look to make sure I didn't break anything.

File: hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs
Comment (MaximilianAlgehed) [reply hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs:122]: Nothing has changed that should influence `parseQuickCheckResult`. It's actually just that in the recent version we decided to include labels that had not been covered in the result. Consequently, the `0% is 5` shows up here. The semantics are the same as always, if there is only one short category it will be printed in brackets with a full stop after `passed NNN tests`. So it _should_ be ok!

Comment (sol) [reply hspec-core/test/Test/Hspec/Core/QuickCheck/UtilSpec.hs:122]: Great 👍

**Review by sol — APPROVED** [review]
Looks great! Thanks a lot  🙏

**Pull request merged**
